### PR TITLE
Image Classification with Swin Transformers

### PR DESCRIPTION
**Pull request by Rishit-dagli**
This example implements [Swin Transformer: Hierarchical Vision Transformer using Shifted Windows](https://arxiv.org/abs/2103.14030) paper by Liu et al. for image classification, and demonstrates it on the [CIFAR-100 dataset](https://www.cs.toronto.edu/~kriz/cifar.html).

Swin Transformer (**S**hifted **Wi**ndow) capably serves as a general-purpose backbone for computer vision. Swin Transformer is a hierarchical Transformer whose representation is computed with shifted windows. The shifted windowing scheme brings greater efficiency by limiting self-attention computation to non-overlapping local windows while also allowing for cross-window connection. This architecture also has the flexibility to model at various scales and has linear computational complexity with respect to image size.

This same example could also be tried out in this Colab Notebook: [Open In Colab](https://colab.research.google.com/drive/1VeYDiDGpG33LcK34faCCDYRebwbUyqsZ?usp=sharing).

> Note: I only train this for 40 epochs (~30 minutes in Google Colab free version) in this example however there are no signs of overfitting at all, so we could probably increase the number of epochs in this example too. As can be seen in this graph.

image

---

This code takes inspiration from the official PyTorch and TensorFlow implementations.

### CODE REVIEW

**Comment by Rishit-dagli** [comment]
It seems the black formatting CI is not failing for this example, [see logs](https://github.com/keras-team/keras-io/pull/625/checks?check_run_id=3532633935#step:4:33):

> would reformat /home/runner/work/keras-io/keras-io/examples/vision/depth_estimation.py
Oh no! 💥 💔 💥
1 file would be reformatted, 102 files would be left unchanged.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! This is a great topic for a code example. I found the code to be very well structured and very high quality -- good job!

File: examples/vision/swin_transformers.py [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+plt.show()
+
+"""
+Let's see final results of the training on CIFAR-10
+"""
+
+loss, accuracy, top_5_accuracy = model.evaluate(x_test, y_test)
+print(f"Test loss: {round(loss, 2)}")
+print(f"Test accuracy: {round(accuracy * 100, 2)}%")
+print(f"Test top 5 accuracy: {round(top_5_accuracy * 100, 2)}%")
```
[the remark this answers is not in the document]
Comment (Rishit-dagli) [reply examples/vision/swin_transformers.py:521]: Sure, I have updated the example to use CIFAR-100 instead

File: examples/vision/swin_transformers.py [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+plt.grid()
+plt.show()
+
+"""
+Let's see final results of the training on CIFAR-10
+"""
+
+loss, accuracy, top_5_accuracy = model.evaluate(x_test, y_test)
+print(f"Test loss: {round(loss, 2)}")
+print(f"Test accuracy: {round(accuracy * 100, 2)}%")
```
[the remark this answers is not in the document]
Comment (Rishit-dagli) [reply examples/vision/swin_transformers.py:520]: That makes sense, as said in the previous comment I now use CIFAR-100 but as you mention I have also added:
- Increased dropout by a minuscule
- 2 Data Augmentation layers, we could most certainly add more and improve performance but I wanted to keep the training time in Colab ~30 minutes which I believe would not be super long
- Increased the number of epochs, but made other changes to compensate for the training time to be almost the same

This gets us to 75% top-5 accuracy on CIFAR-100 with no signs of overfitting so the readers could probably just increase the number of epochs. Do you think this would suffice?

File: examples/vision/swin_transformers.py [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+        self.num_patch = num_patch
+        self.proj = layers.Dense(embed_dim)
+        self.pos_embed = layers.Embedding(input_dim=num_patch, output_dim=embed_dim)
+
+    def call(self, patch):
+        pos = tf.range(start=0, limit=self.num_patch, delta=1)
+        return self.proj(patch) + self.pos_embed(pos)
+
+
+class PatchMerging(tf.keras.layers.Layer):
```
[the remark this answers is not in the document]
Comment (Rishit-dagli) [reply examples/vision/swin_transformers.py:423]: Ah, yes, fixed this

File: examples/vision/swin_transformers.py [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+us to ~96.3% top-5 accuracy within just 20 epochs without any signs of overfitting 
+as well as seen in above graph. This means we can train this network for longer 
+(perhaps with a bit more regularization) and may obtain even better performance. 
+The authors present a top-1 accuracy of 87.3% on ImageNet. The authors also present 
+a number of experiments to study how input sizes, optimizers etc. affect the final 
+performance of this model. The authors further present using this model for object 
+detection and image segmentation as well. You are strongly advised to check out 
+the [original paper](https://arxiv.org/abs/2103.14030) as well.
+
+This example takes inspiration from the official PyTorch and TensorFlow implementations.
```
[the remark this answers is not in the document]
Comment (Rishit-dagli) [reply examples/vision/swin_transformers.py]: Added them

File: examples/vision/swin_transformers.py [hunk 5]
```diff
@@ -0,0 +1,544 @@
...
+complexity with respect to the number of tokens. Here as the [original paper](https://arxiv.org/abs/2103.14030) 
+suggests we compute self-attention within local windows, in a non-overlapping manner. 
+Global self attention introduces quadratic computational complexity to the patch 
+number whereas window based self-attention would be linear and easily scalable.
+"""
+
+
+class WindowAttention(tf.keras.layers.Layer):
+    def __init__(self, dim, window_size, num_heads, qkv_bias=True, drop_rate=0.0):
+        super(WindowAttention, self).__init__()
```
[the remark this answers is not in the document]
Comment (Rishit-dagli) [reply examples/vision/swin_transformers.py]: Sure thing, added this in all places in the example.

File: examples/vision/swin_transformers.py [hunk 6]
```diff
@@ -0,0 +1,544 @@
...
+a token and all other tokens are computed. The global computation leads to quadratic 
+complexity with respect to the number of tokens. Here as the [original paper](https://arxiv.org/abs/2103.14030) 
+suggests we compute self-attention within local windows, in a non-overlapping manner. 
+Global self attention introduces quadratic computational complexity to the patch 
+number whereas window based self-attention would be linear and easily scalable.
+"""
+
+
+class WindowAttention(tf.keras.layers.Layer):
+    def __init__(self, dim, window_size, num_heads, qkv_bias=True, drop_rate=0.0):
```
[the remark this answers is not in the document]
Comment (Rishit-dagli) [reply examples/vision/swin_transformers.py]: Fixed this in all places

File: examples/vision/swin_transformers.py [hunk 7]
```diff
@@ -0,0 +1,544 @@
...
+        relative_coords[:, :, 1] += self.window_size[1] - 1
+        relative_coords[:, :, 0] *= 2 * self.window_size[1] - 1
+        relative_position_index = relative_coords.sum(-1)
+
+        self.relative_position_index = tf.Variable(
+            initial_value=tf.convert_to_tensor(relative_position_index), trainable=False
+        )
+
+    def call(self, x, mask=None):
+        _, N, C = x.get_shape().as_list()
```
[the remark this answers is not in the document]
Comment (Rishit-dagli) [reply examples/vision/swin_transformers.py]: Sure thing. Added fully spelled out variables in all parts.

File: examples/vision/swin_transformers.py [hunk 8]
```diff
@@ -0,0 +1,544 @@
...
+        )
+        self.drop_path = DropPath(drop_rate)
+        self.norm2 = layers.LayerNormalization(epsilon=1e-5)
+        self.mlp = Mlp([num_mlp, dim], drop=drop_rate)
+        if min(self.num_patch) < self.window_size:
+            self.shift_size = 0
+            self.window_size = min(self.num_patch)
+
+    def build(self, input_shape):
+        if self.shift_size > 0:
```
[the remark this answers is not in the document]
Comment (Rishit-dagli) [reply examples/vision/swin_transformers.py]: AH, got it

File: examples/vision/swin_transformers.py [hunk 9]
```diff
@@ -0,0 +1,544 @@
...
+
+
+class DropPath(layers.Layer):
+    def __init__(self, drop_prob=None, **kwargs):
+        super(DropPath, self).__init__(**kwargs)
+        self.drop_prob = drop_prob
+
+    def call(self, x):
+        input_shape = tf.shape(x)
+        batch_num = input_shape[0]
```
[the remark this answers is not in the document]
Comment (Rishit-dagli) [reply examples/vision/swin_transformers.py]: Fixed

File: examples/vision/swin_transformers.py [hunk 10]
```diff
@@ -0,0 +1,544 @@
...
+Usually Transformers conduct global self attention, where the relationships between 
+a token and all other tokens are computed. The global computation leads to quadratic 
+complexity with respect to the number of tokens. Here as the [original paper](https://arxiv.org/abs/2103.14030) 
+suggests we compute self-attention within local windows, in a non-overlapping manner. 
+Global self attention introduces quadratic computational complexity to the patch 
+number whereas window based self-attention would be linear and easily scalable.
+"""
+
+
+class WindowAttention(tf.keras.layers.Layer):
```
[the remark this answers is not in the document]
Comment (Rishit-dagli) [reply examples/vision/swin_transformers.py:176]: Fixed

File: examples/vision/swin_transformers.py [hunk 11]
```diff
@@ -0,0 +1,544 @@
...
+
+"""
+## MLP layer
+
+We will now create a simple multi-layered perceptron with 2 Dense and 2 Dropout layers.
+"""
+
+
+class Mlp(layers.Layer):
+    def __init__(self, filter_num, drop=0.0, **kwargs):
```
[the remark this answers is not in the document]
Comment (Rishit-dagli) [reply examples/vision/swin_transformers.py:148]: Fixed in response to your earlier comment of fully-spelled out variable names

**Comment by Rishit-dagli** [comment]
Thanks so much, @fchollet for the insightful reviews. I have updated the example likewise incorporating all your reviews. Apart from the reviews, I also noticed the need to restructure very few comments and have done that too.

File: examples/vision/swin_transformers.py [hunk 12]
```diff
@@ -0,0 +1,557 @@
...
+## Setup
+"""
+
+import matplotlib.pyplot as plt
+import numpy as np
+import tensorflow as tf
+import tensorflow_addons as tfa
+from tensorflow import keras
+from tensorflow.keras import layers
+from tensorflow.python.keras.layers.pooling import GlobalAveragePooling1D
```
Comment (fchollet) [inline examples/vision/swin_transformers.py:39]: This is a private API, instead use `layers.GlobalAveragePooling1D`

File: examples/vision/swin_transformers.py [hunk 13]
```diff
@@ -0,0 +1,557 @@
...
+## Helper Functions
+
+We will now create two helper functions which can help us get a sequence of 
+patches from the image, allow us to merge patches to spatial frames and dropout.
+"""
+
+
+def window_partition(x, window_size):
+    _, height, width, channels = x.shape
+    patch_num_H = height // window_size
```
Comment (fchollet) [inline examples/vision/swin_transformers.py:106]: Style: use lowercase (snake case) variable names everywhere

File: examples/vision/swin_transformers.py [hunk 14]
```diff
@@ -0,0 +1,557 @@
...
+
+
+"""
+## MLP layer
+
+We will now create a simple multi-layered perceptron with 2 Dense and 2 Dropout layers.
+"""
+
+
+class Mlp(layers.Layer):
```
Comment (fchollet) [inline examples/vision/swin_transformers.py:151]: This class is equivalent to 

```
keras.Sequential(
    [
            layers.Dense(filter_num[0]),
            layers.Activation(keras.activations.gelu),
            layers.Dropout(dropout_rate),
            layers.Dense(filter_num[1]),
            layers.Dropout(dropout_rate),
    ])
```

No need to wrap it.

Since it is only used once, I would recommend in-line the Sequential definition where the model is used.

File: examples/vision/swin_transformers.py [hunk 15]
```diff
@@ -0,0 +1,557 @@
...
+        self.attn = WindowAttention(
+            dim,
+            window_size=(self.window_size, self.window_size),
+            num_heads=num_heads,
+            qkv_bias=qkv_bias,
+            dropout_rate=dropout_rate,
+        )
+        self.drop_path = DropPath(dropout_rate)
+        self.norm2 = layers.LayerNormalization(epsilon=1e-5)
+        self.mlp = Mlp([num_mlp, dim], drop=dropout_rate)
```
Comment (fchollet) [inline examples/vision/swin_transformers.py:309]: that is to say, here

File: examples/vision/swin_transformers.py [hunk 16]
```diff
@@ -0,0 +1,557 @@
...
+""""
+### Extract and Embed patches
+
+We will first create 3 layers to help us extract, embed and merge patches from the 
+images on top of which we will later use the Swin Transfromer class we built.
+"""
+
+
+class PatchExtract(layers.Layer):
+    def __init__(self, patch_size):
```
Comment (fchollet) [inline examples/vision/swin_transformers.py:404]: Add `**kwargs` support

File: examples/vision/swin_transformers.py [hunk 17]
```diff
@@ -0,0 +1,557 @@
...
+            rates=(1, 1, 1, 1),
+            padding="VALID",
+        )
+        patch_dim = patches.shape[-1]
+        patch_num = patches.shape[1]
+        return tf.reshape(patches, (batch_size, patch_num * patch_num, patch_dim))
+
+
+class PatchEmbedding(layers.Layer):
+    def __init__(self, num_patch, embed_dim):
```
Comment (fchollet) [inline examples/vision/swin_transformers.py:424]: Likewise

File: examples/vision/swin_transformers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fchollet) [reply examples/vision/swin_transformers.py:535]: It's fine to train the model for fewer epochs than needed in order to keep the runtime short. See instructions here: https://github.com/keras-team/keras-io/blob/master/contributor_guide.md#demonstrate-the-full-power-of-your-model-but-keep-the-run-time-short

What do you get when you train the model to convergence? A reasonable accuracy (not top 5) on CIFAR 100 would be higher than 70%, for the full model (which may be trained on as much resources as you want)

**Review by fchollet — COMMENTED** [review]
Thanks for the update!

File: examples/vision/swin_transformers.py (hunk 12 above)
Comment (Rishit-dagli) [reply examples/vision/swin_transformers.py]: Got it, I always used to wonder what `.python.keras.*` is for, thanks

File: examples/vision/swin_transformers.py (hunk 13 above)
Comment (Rishit-dagli) [reply examples/vision/swin_transformers.py]: Thanks for pointing this out, might  have missed this one in an earlier commit

File: examples/vision/swin_transformers.py (hunk 14 above)
Comment (Rishit-dagli) [reply examples/vision/swin_transformers.py]: That makes sense but we get the values of `filter_num` only while inside the Swin Transformer block so I have defined the MLP inside the Swin Transformer class

File: examples/vision/swin_transformers.py (hunk 16 above)
Comment (Rishit-dagli) [reply examples/vision/swin_transformers.py]: Ah, might have missed this class

File: examples/vision/swin_transformers.py (hunk 17 above)
Comment (Rishit-dagli) [reply examples/vision/swin_transformers.py]: Ah, might have missed this class

File: examples/vision/swin_transformers.py (hunk 15 above)
Comment (Rishit-dagli) [reply examples/vision/swin_transformers.py]: Got it, put the MLP definition inline in this code block

**Comment by Rishit-dagli** [comment]
> It's fine to train the model for fewer epochs than needed in order to keep the runtime short. See instructions here: https://github.com/keras-team/keras-io/blob/master/contributor_guide.md#demonstrate-the-full-power-of-your-model-but-keep-the-run-time-short
What do you get when you train the model to convergence? A reasonable accuracy (not top 5) on CIFAR 100 would be higher than 70%, for the full model (which may be trained on as much resources as you want)

Ah, got it. Thanks for sharing this section of the contributing guide. I have made the edits accordingly. I trained the model for 150 epochs with slightly higher dropout and more embeddings dimensions which gets the test accuracy to just above 70% (~72%) which as you mention should work. Also, I have updated the comments to reflect this.

**Comment by Rishit-dagli** [comment]
Thanks for the great reviews 👍 , I have implemented them

**Review by fchollet — COMMENTED** [review]
Awesome! I pushed some light copyedits, please pull them first. Please add the generated files afterwards, I believe we're ready to publish!

**Comment by Rishit-dagli** [comment]
> Awesome! I pushed some light copyedits, please pull them first. Please add the generated files afterwards, I believe we're ready to publish!

Thanks for making the copyedits (and fixing multiple typos by me 😅)

**Review by fchollet — APPROVED** [review]
Thank you! Please add the generated files.

**Comment by Rishit-dagli** [comment]
@fchollet Thanks! Added the generated files.

**Comment by fchollet** [comment]
Thank you for the great contribution! 👍
